### ISSUE TITLE
JDBC Memory leak on CentOS 7.7

### ISSUE THREAD

**Issue by Cinzq4615** [duckdb/duckdb#3805]
#### What happens?
Memory leak on CentOS 7.7,window is ok.
There is a memory leak when executing multiple queries  on CentOS 7.7

#### To Reproduce
We've written a minimal reproduction and shared it as [a GitHub repository](https://github.com/Cinzq4615/duckdb-leak-reproduction.git). The reproduction runs for a specified period of time.

#### Environment (please complete the following information):
 - OS: CentOS 7.7
 - DuckDB Version: 0.3.3
 - DuckDB Client: jdbc
- [x] ] **Have you tried this on the latest `master` branch?**
* **Python**: `pip install duckdb --upgrade --pre`
* **R**: `install.packages("https://github.com/duckdb/duckdb/releases/download/master-builds/duckdb_r_src.tar.gz", repos = NULL)`
* **Other Platforms**: You can find binaries [here](https://github.com/duckdb/duckdb/releases/tag/master-builds) or compile from source.

- [x] **Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?**

**Comment by Mytherin** [comment]
Thanks for the detailed report!

I have tried to reproduce this using the C++ API, but could not reproduce the issue there using the following script:

```cpp

TEST_CASE("Test leak check", "[api]") {
    string DATABASE_SIZE_QUERY = "pragma database_size";
    string TABLE_COUNT_QUERY = "select count(id) as count from mydb.test";


    string SCHEMA_CREATE_SQL = "CREATE SCHEMA IF NOT EXISTS mydb;";

    string TABEL_CREATE_SQL = "CREATE TABLE IF NOT EXISTS mydb.test(id INTEGER, name VARCHAR);";

    string TABEL_DROP_SQL = "DROP TABLE IF EXISTS mydb.test;";
	while(true) {
		auto db = make_unique<DuckDB>(nullptr);
		auto conn = make_unique<Connection>(*db);

		conn->Query("PRAGMA threads=1");
		conn->Query(SCHEMA_CREATE_SQL);
		conn->Query(TABEL_CREATE_SQL);

		for(idx_t i = 0; i < 1000; i++) {
			Appender appender(*conn, "mydb", "test");
			for(idx_t j = 0; j < 100000; j++) {
				appender.BeginRow();
				appender.Append(j);
				appender.Append("name-" + to_string(j));
				appender.EndRow();
			}
			appender.Close();
		}
		conn->Query(TABLE_COUNT_QUERY)->Print();
		conn->Query(TABEL_DROP_SQL);

		conn.reset();
		db.reset();
	}
}
```

I get the following RSS size over a ~20 minute run, with the following Python script running to measure this:

```py
import subprocess
import time

f = open('result.csv', 'w+')

while True:
	time.sleep(1)
	proc = subprocess.Popen('ps -o rss= -p 85917'.split(' '), stdout=subprocess.PIPE)
	res = proc.stdout.read().decode('utf8').strip()

	f.write(res)
	f.write('\n')
	f.flush()
```

<img src="https://i.imgur.com/rE0KzEz.png">

This is most likely a problem specific to the JDBC driver, either in the JNI or in the JDBC driver itself.

**Comment by Cinzq4615** [comment]
Thanks，there is another question when using the C++ API， running with a persistent database,drop table ,memory can not release.

**Comment by Mytherin** [comment]
It appears to me that you are using DuckDB in an enterprise setup. May I ask what your affiliation is/who you work for?

**Comment by Cinzq4615** [comment]
> It appears to me that you are using DuckDB in an enterprise setup. May I ask what your affiliation is/who you work for?
I work for wind.

**Comment by Cinzq4615** [comment]
I get the following memory allocate,JDBC running to measure this:
```
        while (true) {
            try (Statement statement = connection.createStatement();ResultSet rs = statement.executeQuery(String.format(TABEL_QUERY_SQL,random.nextInt(10000)))) {
                if(rs.next()){
                    System.out.println(rs.getString(2));
                }
            }
        }
```
![windexport_1655272580773](https://user-images.githubusercontent.com/19431346/173757863-34b7f96c-27e6-4a00-9e12-220198747af2.jpg)

**Comment by Cinzq4615** [comment]
Each use of  Statement is wrapped in a try-with-resources block. So they should be closed automatically.

**Comment by hannes** [comment]
"should"

**Comment by Cinzq4615** [comment]
using the following code
```
public class Main {
   // How long the test should run.
   private static final Duration DEFAULT_TEST_DURATION = Duration.ofHours(1);
   // How long to wait for the thread pool to shut down when quitting.
   private static final Duration SHUTDOWN_TIMEOUT = Duration.ofSeconds(5);
   //private static final String DATABASE_SIZE_QUERY = "pragma database_size";
   private static final String TABLE_COUNT_QUERY = "select count(id) as tmp_count from mydb.tmp;";
   private static final String TABLE_SCHEMA_SQL = "CREATE SCHEMA IF NOT EXISTS mydb";
   private static final String TABLE_CREATE_SQL = "CREATE TABLE IF NOT EXISTS mydb.tmp (\n" +
           "\tid INTEGER NOT NULL,\n" +
           "\tcolumn_1 VARCHAR,\n" +
           "\tcolumn_2 VARCHAR,\n" +
           "\tcolumn_3 VARCHAR,\n" +
           "\tcolumn_4 VARCHAR,\n" +
           "\tcolumn_5 VARCHAR,\n" +
           "\tcolumn_6 VARCHAR,\n" +
           "\tcolumn_7 VARCHAR,\n" +
           "\tcolumn_8 VARCHAR,\n" +
           "\tcolumn_9 VARCHAR,\n" +
           "\tcolumn_10 VARCHAR,\n" +
           "\tcolumn_11 VARCHAR,\n" +
           "\tcolumn_12 VARCHAR,\n" +
           "\tcolumn_13 VARCHAR,\n" +
           "\tcolumn_14 VARCHAR,\n" +
           "\tcolumn_15 VARCHAR,\n" +
           "\tcolumn_16 VARCHAR,\n" +
           "\tcolumn_17 VARCHAR,\n" +
           "\tcolumn_18 VARCHAR,\n" +
           "\tcolumn_19 VARCHAR,\n" +
           "\tcolumn_20 VARCHAR,\n" +
           "\tcolumn_21 VARCHAR,\n" +
           "\tcolumn_22 VARCHAR,\n" +
           "\tcolumn_23 VARCHAR,\n" +
           "\tcolumn_24 VARCHAR,\n" +
           "\tcolumn_25 VARCHAR,\n" +
           "\tcolumn_26 VARCHAR,\n" +
           "\tcolumn_27 VARCHAR,\n" +
           "\tcolumn_28 VARCHAR,\n" +
           "\tcolumn_29 VARCHAR,\n" +
           "\tcolumn_30 VARCHAR,\n" +
           "\tcolumn_31 VARCHAR,\n" +
           "\tcolumn_32 VARCHAR,\n" +
           "\tcolumn_33 VARCHAR,\n" +
           "\tcolumn_34 VARCHAR,\n" +
           "\tcolumn_35 VARCHAR,\n" +
           "\tcolumn_36 VARCHAR,\n" +
           "\tcolumn_37 VARCHAR,\n" +
           "\tcolumn_38 VARCHAR,\n" +
           "\tcolumn_39 VARCHAR,\n" +
           "\tcolumn_40 VARCHAR,\n" +
           "\tcolumn_41 VARCHAR,\n" +
           "\tcolumn_42 VARCHAR,\n" +
           "\tcolumn_43 VARCHAR,\n" +
           "\tcolumn_44 VARCHAR,\n" +
           "\tcolumn_45 VARCHAR,\n" +
           "\tcolumn_46 VARCHAR,\n" +
           "\tcolumn_47 VARCHAR,\n" +
           "\tcolumn_48 VARCHAR,\n" +
           "\tcolumn_49 VARCHAR,\n" +
           "\tcolumn_50 VARCHAR\n" +
           ");";
   private static final String DATA_DROP_SQL = "DROP TABLE mydb.tmp;";
   private static final String DATA_SELECT_SQL = "SELECT * FROM mydb.tmp WHERE id = 5;";
public static void main(String[] args) throws Exception {
       // Determine how long the test is going to run.
       Duration testDuration = getTestDuration(args);
       // Do the date math to figure out when the test should shut down.
       ZonedDateTime startingAt = ZonedDateTime.now();
       ZonedDateTime endingAt = startingAt.plus(testDuration);
       DuckDBConnection connection = (DuckDBConnection) DriverManager.getConnection("jdbc:duckdb::memory:");
       try (Statement statement = connection.createStatement()) {
           statement.execute("PRAGMA threads=3;");
       }
       try (Statement statement = connection.createStatement()) {
           statement.execute("PRAGMA memory_limit='0.5GB';");
       }
       //create schema
       try (Statement statement = connection.createStatement()) {
           statement.execute(TABLE_SCHEMA_SQL);
       }
       //create table
       try (Statement statement = connection.createStatement()) {
           statement.execute(TABLE_CREATE_SQL);
       }
       DuckDBAppender appender;
       appender = connection.createAppender("mydb", "tmp");
       for (int i = 0; i < 10; i++) {
           appender.beginRow();
           appender.append(i);
           for (int j = 1; j < 51; j++) {
               appender.append("column-" + j);
           }
           appender.endRow();
       }
       appender.close();
 
       DuckDBConnection conn = null;
       Statement statement = null;
       ResultSet resultSet = null;
       while (ZonedDateTime.now().isBefore(endingAt)) {
           try {
               conn = (DuckDBConnection) connection.duplicate();
               statement = conn.createStatement();
               resultSet = statement.executeQuery(DATA_SELECT_SQL);
               if (resultSet.next()) {
                   System.out.println(resultSet.getString(2));
               }
           } finally {
               if (resultSet != null) {
                   resultSet.close();
               }
               if (statement != null) {
                   statement.close();
               }
               if (conn != null) {
                   conn.close();
               }
           }
       }
       DuckDBDatabase db = connection.getDatabase();
       connection.close();
       db.shutdown();
   }
   /**
    * Parse the test duration from the command line arguments. If no duration was supplied
    * then use the default. Otherwise parse the duration using {@link Duration}.parse(String).
    *
    * @param args The command line arguments to parse.
    * @return The parsed duration or the default if no duration was given.
    */
   private static Duration getTestDuration(String[] args) {
       if (args.length == 0) {
           return DEFAULT_TEST_DURATION;
       }
       return Duration.parse(args[0]);
   }
}
```

**Comment by rkennedy-mode** [comment]
> Each use of Statement is wrapped in a try-with-resources block. So they should be closed automatically.

To @hannes' remark of "should"…any resources in a Java try-with-resources is guaranteed to be closed at the end of the try block: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

> The try-with-resources statement is a try statement that declares one or more resources. A resource is an object that must be closed after the program is finished with it. The try-with-resources statement ensures that each resource is closed at the end of the statement.

**Comment by Jens-H** [comment]
#3610 was a very similar problem that got solved recently. So just to make sure I got the facts right: 

1 this happens only on Linux/Centos and not on windows
> Memory leak on CentOS 7.7,window is ok.

2 with an actual build from master
3 and this loop is causing it

> while (true) {
>             try (Statement statement = connection.createStatement();ResultSet rs = statement.executeQuery(String.format(TABEL_QUERY_SQL,random.nextInt(10000)))) {
>                 if(rs.next()){
>                     System.out.println(rs.getString(2));
>                 }
>             }
>         }

Did I miss anything?

**Comment by Cinzq4615** [comment]
Had made a mistake before ，window is not ok. I sure， build from master.

**Comment by Jens-H** [comment]
@Mytherin 
The JDBC driver does not use Query. Everything goes through Prepare and Execute.
 
> ```c++
> conn->Query(TABEL_DROP_SQL);
> ```

I ran this example (that I feel is close to the JDBC calls) and each query uses almost a 200 Bytes of extra memory.
```c++
#include "duckdb.hpp"
#include <iostream>

using namespace duckdb;

int main() {
	DuckDB db(nullptr);
	/* Connection con(db); */
    auto con = make_unique<Connection>(db);

	con->Query("create table t (id int8, texty varchar)");
	con->Query("insert into t values (1, 'uiae')");

    int i = 0;
    long total = 0;

    while (true)
    {
        auto stmt_ref = con->Prepare("select texty from t where id = 1;");
        vector<Value> duckdb_params;
        auto res_ref = stmt_ref->Execute(duckdb_params, false);
        
        i++;
        total++;

        if (i == 1000)
        {
            std::cout << "Total: " <<  total << std::endl;
            i = 0;
        }
    }
}
```

Hopefully someone can confirm this or point to a problem with the example.

**Comment by Mytherin** [comment]
> @Mytherin The JDBC driver does not use Query. Everything goes through Prepare and Execute.
> 
> > ```c++
> > conn->Query(TABEL_DROP_SQL);
> > ```
> 
> I ran this example (that I feel is close to the JDBC calls) and each query uses almost a 200 Bytes of extra memory.
> 
> ```c++
> #include "duckdb.hpp"
> #include <iostream>
> 
> using namespace duckdb;
> 
> int main() {
> 	DuckDB db(nullptr);
> 	/* Connection con(db); */
>     auto con = make_unique<Connection>(db);
> 
> 	con->Query("create table t (id int8, texty varchar)");
> 	con->Query("insert into t values (1, 'uiae')");
> 
>     int i = 0;
>     long total = 0;
> 
>     while (true)
>     {
>         auto stmt_ref = con->Prepare("select texty from t where id = 1;");
>         vector<Value> duckdb_params;
>         auto res_ref = stmt_ref->Execute(duckdb_params, false);
>         
>         i++;
>         total++;
> 
>         if (i == 1000)
>         {
>             std::cout << "Total: " <<  total << std::endl;
>             i = 0;
>         }
>     }
> }
> ```
> 
> Hopefully someone can confirm this or point to a problem with the example.

Thanks! Indeed I can confirm this does seem to leak memory. I will investigate.

**Comment by Mytherin** [comment]
Should be fixed in #3894, the problem is that to-be-freed entries accumulate in the buffer manager (to the tune of ~some bytes) which are then never looked at again if the system is never in a situation where it is low on memory. The problem with my test case was that I actually closed the database in between runs, which frees up all the accumulated entries in the buffer manager.
